File: WdtcCore/src/main/java/org/wdt/wdtc/core/game/GetGameNeedLibraryFile.java
package org.wdt.wdtc.core.game;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import lombok.Getter;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class GetGameNeedLibraryFile {
  private final Launcher launcher;

  public GetGameNeedLibraryFile(Launcher launcher) {
    this.launcher = launcher;
  }

  public List<LibraryFile> getFileList() throws IOException {
    List<LibraryFile> FileList = new ArrayList<>();
    GameVersionJsonObject VersionJsonObject = launcher.getGameVersionJsonObject();
    for (LibraryObject LibraryObject : VersionJsonObject.getLibraries()) {
      if (LibraryObject.getNatives() != null) {
        JsonObject NativesJson = LibraryObject.getNatives();
        if (NativesJson.has("windows")) {
          FileList.add(new LibraryFile(LibraryObject, true));
        }
      } else {
        JsonArray rules = LibraryObject.getRules();
        if (rules != null) {
          JsonObject ActionObject = rules.get(rules.size() - 1).getAsJsonObject();
          String action = ActionObject.get("action").getAsString();
          String OsName = ActionObject.getAsJsonObject("os").get("name").getAsString();
          if (Objects.equals(action, "disallow") && Objects.equals(OsName, "osx")) {
            FileList.add(new LibraryFile(LibraryObject));
          } else if (Objects.equals(action, "allow") && Objects.equals(OsName, "windows")) {
            FileList.add(new LibraryFile(LibraryObject));
          }
        } else {
          FileList.add(new LibraryFile(LibraryObject));
        }
      }
    }
    return FileList;
  }

  @Getter
  public static class LibraryFile {
    private final LibraryObject libraryObject;
    private final boolean NativesLibrary;

    public LibraryFile(LibraryObject libraryObject, boolean nativesLibrary) {
      this.libraryObject = libraryObject;
      NativesLibrary = nativesLibrary;
    }

    public LibraryFile(LibraryObject libraryObject) {
      this(libraryObject, false);
    }

  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/game/GetGameNeedLibraryFile.kt
package org.wdt.wdtc.core.game

import org.wdt.utils.gson.getJsonObject
import org.wdt.utils.gson.getString
import java.io.IOException

class GetGameNeedLibraryFile(private val launcher: Launcher) {
  @get:Throws(IOException::class)
  val libraryList: List<LibraryFile>
    get() {
      val fileList: MutableList<LibraryFile> = ArrayList()
      val versionJsonObject = launcher.gameVersionJsonObject
      for (libraryObject in versionJsonObject.libraries!!) {
        if (libraryObject.natives != null) {
          val nativesJson = libraryObject.natives!!
          if (nativesJson.has("windows")) {
            fileList.add(LibraryFile(libraryObject, true))
          }
        } else {
          val rules = libraryObject.rules
          if (rules != null) {
            val actionObject = rules.getJsonObject(rules.size() - 1)
            val action = actionObject.getString("action")
            val osName = actionObject.getJsonObject("os").getString("name")
            if (action == "disallow" && osName == "osx") {
              fileList.add(LibraryFile(libraryObject))
            } else if (action == "allow" && osName == "windows") {
              fileList.add(LibraryFile(libraryObject))
            }
          } else {
            fileList.add(LibraryFile(libraryObject))
          }
        }
      }
      return fileList
    }

  class LibraryFile @JvmOverloads constructor(
    var libraryObject: LibraryObject,
    var nativesLibrary: Boolean = false
  )
}
